File: May 2024 GFG SOLUTION/May-19.py

from typing import List

class Solution:
    def findClosest(self, n: int, k: int, arr: List[int]) -> int:
        from bisect import bisect_left
        r = bisect_left(arr, k)
        l = r - 1
        if l < 0:
            return arr[r]
        else:
            diff1 = abs(k - arr[r]) if r < n else float('inf')
            diff2 = abs(k - arr[l])
            if diff1 <= diff2:
                return arr[r]
            return arr[l]


#{ 
 # Driver Code Starts

class IntArray:
    def __init__(self) -> None:
        pass
    def Input(self,n):
        arr=[int(i) for i in input().strip().split()]#array input
        return arr
    def Print(self,arr):
        for i in arr:
            print(i,end=" ")
        print()


if __name__=="__main__":
    t = int(input())
    for _ in range(t):
        
        n = int(input())
        
        
        k = int(input())
        
        
        arr=IntArray().Input(n)
        
        obj = Solution()
        res = obj.findClosest(n, k, arr)
        
        print(res)
        

# } Driver Code Ends

File: May 2024 GFG SOLUTION/May-19.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;


class IntArray
{
    public static int[] input(BufferedReader br, int n) throws IOException
    {
        String[] s = br.readLine().trim().split(" ");
        int[] a = new int[n];
        for(int i = 0; i < n; i++)
            a[i] = Integer.parseInt(s[i]);

        return a;
    }

    public static void print(int[] a)
    {
        for(int e : a)
            System.out.print(e + " ");
        System.out.println();
    }

    public static void print(ArrayList<Integer> a)
    {
        for(int e : a)
            System.out.print(e + " ");
        System.out.println();
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while(t-- > 0){
            
            int n;
            n = Integer.parseInt(br.readLine());
            
            
            int k;
            k = Integer.parseInt(br.readLine());
            
            
            int[] arr = IntArray.input(br, n);
            
            Solution obj = new Solution();
            int res = obj.findClosest(n, k, arr);
            
            System.out.println(res);
            
        }
    }
}

// } Driver Code Ends



class Solution {
    public static int findClosest(int n, int k, int[] arr) {
        int r = Arrays.binarySearch(arr, k);
        if (r < 0) {
            r = -r - 1;
        }
        int l = r - 1;
        if (r >= n) {
            return arr[l];
        }
        if (l < 0) {
            return arr[r];
        } else {
            int diff1 = Math.abs(k - arr[r]);
            int diff2 = Math.abs(k - arr[l]);
            if (diff1 <= diff2) {
                return arr[r];
            }
            return arr[l];
        }
    }
}

        
